Checkout: Add thank you page for the domain only site flow

gziolo opened the pull request:
Implements #11452. ~~Requires D4612-code.~~

This PR:

- Creates `ThankYouCard` from `PlanThankYouCard`.
- Refactors `PlanThankYouCard` to use `ThankYouCard`.
- Uses `ThankYouCard` to display a domain-only thank you page.

### Testing
*`PlanThankYouCard`*
- Make sure https://wpcalypso.wordpress.com/devdocs/blocks/plan-thank-you-card and http://calypso.localhost:3000/devdocs/blocks/plan-thank-you-card looks the same after refactoring. See also image below for comparison.
![screen shot 2017-03-02 at 17 39 49](https://cloud.githubusercontent.com/assets/699132/23517330/4f6d7746-ff70-11e6-8a62-dedddad8d8f1.png)

- Visit `/start` while logged out (`PlanThankYouCard` is only displayed for new users).
- Select a non-free plan on the plans page.
- Proceed through signup and create a new user.
- Assert that you land in checkout with the plan you selected in the second step.
- Purchase the plan.
- Assert that you land on a thank you page with `PlanThankYouCard` displayed for the plan you purchased like in the image above.

*New thank you page*
- Visit `/start/domain-first?new=DOMAIN` where `DOMAIN` is a domain like `thisisaniquedomain123.live`.
- Select `Just purchase a domain`.
- Purchase the domain in checkout.
- Assert that you land on a thank you page with `ThankYouCard` displayed for the domain you purchased like:
![screen shot 2017-03-06 at 12 21 00](https://cloud.githubusercontent.com/assets/699132/23608603/72193b5c-026a-11e7-85f7-40907b4be8de.png)


### Review
* [x] Code
* [x] Copy
* [x] Design
* [x] Product

drewblaisdell commented on client/my-sites/upgrades/index.js:
```diff
@@ -234,7 +234,7 @@ module.exports = function() {
 	if ( config.isEnabled( 'upgrades/checkout' ) ) {
 		page(
-			'/checkout/thank-you/no-site/:receiptId',
+			'/checkout/thank-you/no-site/:receiptId?',
```
Thanks for fixing this.

drewblaisdell commented on client/my-sites/upgrades/checkout-thank-you/index.jsx:
```diff
@@ -229,6 +229,17 @@ const CheckoutThankYou = React.createClass( {
 			);
 		}
 
+		if ( this.props.domainOnlySiteFlow && ! failedPurchases ) {
```
Empty arrays are still falsey so we need to check the length here.

gziolo replied:
I had this locally:)

```js
if ( this.props.domainOnlySiteFlow && failedPurchases.length === 0 && purchases.some( isDomainRegistration ) ) {
```

drewblaisdell commented on client/blocks/plan-thank-you-card/index.jsx:
```diff
@@ -16,76 +15,43 @@ import QuerySites from 'components/data/query-sites';
-						{ translate( 'Visit Your Site' ) }
-					</a>
-				</div>
+			<div className="plan-thank-you-card__heading">
```
I _think_ these styles should belong in `ThankYouCard` instead of `PlanThankYouCard` because we want to use them in every (currently, both) version of the thank you card.

gziolo replied:
I'm perfectly fine with this approach. This was my first idea, but I quickly got distracted by the power of `children` :)

What you proposed fits what we need in this particular approach and removes code duplication. If it turns out we need something different inside body we can always extract this content and create another component and get back to the concept of using `children`.

a8ci18n commented on client/my-sites/upgrades/checkout-thank-you/index.jsx:
```diff
@@ -229,6 +232,27 @@ const CheckoutThankYou = React.createClass( {
+						heading={ this.translate( 'Thank you for your purchase!' ) }
+						description={ this.translate( "That looks like a great domain. Now it's time to get it all set up." ) }
+						buttonUrl={ domainManagementEdit( domainName, domainName ) }
+						buttonText={ this.translate( 'Manage Your Domain' ) }
```
Hi! I've found a possible matching string that has already been translated **52** times:
`translate( 'Manage Your Domains' )` *ES Score: 11*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/domain-only-site-thank-you)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

gziolo replied:
@ranh suggested in #11452 to use `Go To Your Domain` instead.

drewblaisdell commented:
@gziolo I think this is ready for review. :)

gziolo commented on client/my-sites/upgrades/checkout-thank-you/index.jsx:
```diff
@@ -194,23 +197,20 @@ const CheckoutThankYou = React.createClass( {
-
 		// this placeholder is using just wp logo here because two possible states do not share a common layout
-		if ( ! purchases && ! failedPurchases && ! this.isGenericReceipt() ) {
+		if ( ! purchases.length && ! failedPurchases.length && ! this.isGenericReceipt() ) {
```
Good idea 👍

stephanethomas replied:
That's personal preference but I like using [`isEmpty()`](https://lodash.com/docs#isEmpty) in those cases since it makes the code reads like a sentence.

gziolo commented on client/my-sites/upgrades/checkout-thank-you/index.jsx:
```diff
@@ -229,6 +232,27 @@ const CheckoutThankYou = React.createClass( {
+
+					<ThankYouCard
+						name={ this.translate( 'Domain Registration' ) }
+						price={ purchase.displayPrice }
```
What about privacy protection?

gziolo commented:
@drewblaisdell thanks for finishing this PR. Everything looks awesome.

I had only two comments, one related to the different copy for the button. We also need to double check if privacy protection cost is included in the price displayed on this page.

gziolo commented:
This is how the page actually works. On the left I added view from my inbox to show the difference between both prices:

![screen shot 2017-03-03 at 17 37 07](https://cloud.githubusercontent.com/assets/699132/23559699/231e7a1e-0038-11e7-9371-49c6eb6ada38.png)

mikedang commented:
Copy looks good!

danhauk commented:
The thank you message shows as expected and looks good. I can also confirm that the privacy protection cost was not added to the total cost displayed for me.

With the change of the button copy, I think it would be better if the button took the user to `/domains/manage/:name` instead of `/domains/manage/:name/edit/:name`.

gziolo commented:
We discussed with @ranh on Slack small tweaks to the proposed copy as follows:
- replace `Domain Registration` with the actual domain name
- replace button text `Manage Your Domain` with `Go To Your Domain`
- update link as @danhauk suggested

![screen shot 2017-03-06 at 12 21 00](https://cloud.githubusercontent.com/assets/699132/23608701/e2bbf8a4-026a-11e7-8adc-6a8053a88ef5.png)

gziolo commented:
I added couple of final tweaks and proposed another way to fetch display price. It requires now different backend patch `D4612-code`. I would mark code changes as accepted, but I added another commit. So this needs to be reviewed.

stephanethomas commented:
I applied the patch D4612-code, sandboxed the API, and followed the second set of testing instructions to get to the new `Thank You` page. This page displays correctly except for a tiny detail:

<img width="367" alt="screenshot" src="https://cloud.githubusercontent.com/assets/594356/23617650/091a04f0-028e-11e7-95a1-a68e1fb88cbd.png">

The price indeed doesn't show up. This is where I also get lost. The `/me/transactions` API endpoint indeed didn't return any `display_price`. Even if it was, I'm not sure how it could make its way to the global state tree since it is not passed [in the method](https://github.com/Automattic/wp-calypso/blob/update/domain-only-site-thank-you/client/my-sites/upgrades/checkout/checkout.jsx#L287-L291) that handles the redirect to the `Thank You` page.

drewblaisdell commented:
Thanks @stephanethomas, you diagnosed that issue correctly. I updated the patch/PR to fix it.

gziolo commented:
@drewblaisdell thanks for fixing the issue raised by @stephanethomas 👍 
Code looks good now!

gziolo commented:
@stephanethomas can you try to reproduce the issue you discovered with the updated backend patch and client code? :)

stephanethomas commented on client/my-sites/upgrades/checkout/checkout.jsx:
```diff
@@ -288,10 +282,12 @@ const Checkout = React.createClass( {
 		if ( receipt && receipt.receipt_id ) {
-			const receiptId = receipt.receipt_id;
+			const receiptId = receipt.receipt_id,
+				displayPrice = receipt.display_price;
```
This is where destructuring comes in handy:

```
const { receipt_id: receiptId, display_price: displayPrice } = receipt;
```

stephanethomas commented:
@gziolo: I've just tested it and had the [same issue](https://github.com/Automattic/wp-calypso/pull/11702#issuecomment-284441275) again with the `Thank You` page. Note this time the `/me/transactions` endpoint returned the `display_price` property, but it was equal to zero:

```
{
  "code": 200,
  "headers": [{
    "name": "Content-Type",
    "value": "application\/json"
  }],
  "body": {
    "success": true,
    "error_code": "",
    "error_message": "",
    "receipt_id": "22457622",
    "display_price": "$0.00",
    "purchases": {
      ...
```

drewblaisdell commented:
@stephanethomas There was an issue where we were camelCasing keys that were expected to be raw from the API that I fixed in 52cd24b3c41211ddedd5ffd1b0712c191562a546.

I tested several times with and without the privacy upgrade and never saw a `display_price` of `"$0.00"`. Would you mind trying again and seeing where in the patch this fails for you? :) (We should probably move this discussion there)

gziolo commented:
It might be because @stephanethomas has tested it with credits. I'll double check it today with free credits :) I guess 0 price is sth expected in such case.

gziolo commented:
I tested also the case when paying with card, everything works properly. Product 👍
